app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenProviderImpl.kt

```diff
@@ -0,0 +1,117 @@
+package org.schabi.newpipe.util.potoken
+
+import android.os.Handler
+import android.os.Looper
+import android.util.Log
+import org.schabi.newpipe.App
+import org.schabi.newpipe.BuildConfig
+import org.schabi.newpipe.extractor.NewPipe
+import org.schabi.newpipe.extractor.services.youtube.InnertubeClientRequestInfo
+import org.schabi.newpipe.extractor.services.youtube.PoTokenProvider
+import org.schabi.newpipe.extractor.services.youtube.PoTokenResult
+import org.schabi.newpipe.extractor.services.youtube.YoutubeParsingHelper
+import org.schabi.newpipe.util.DeviceUtils
+
+object PoTokenProviderImpl : PoTokenProvider {
+    val TAG = PoTokenProviderImpl::class.simpleName
+    private val webViewSupported by lazy { DeviceUtils.supportsWebView() }
+
+    private object WebPoTokenGenLock
+    private var webPoTokenVisitorData: String? = null
+    private var webPoTokenStreamingPot: String? = null
+    private var webPoTokenGenerator: PoTokenGenerator? = null
+
+    override fun getWebClientPoToken(videoId: String): PoTokenResult? {
+        if (!webViewSupported) {
+            return null
+        }
+
+        return getWebClientPoToken(videoId = videoId, forceRecreate = false)
+    }
+
+    /**
+     * @param forceRecreate whether to force the recreation of [webPoTokenGenerator], to be used in
+     * case the current [webPoTokenGenerator] threw an error last time
+     * [PoTokenGenerator.generatePoToken] was called
+     */
+    private fun getWebClientPoToken(videoId: String, forceRecreate: Boolean): PoTokenResult {
+        // just a helper class since Kotlin does not have builtin support for 4-tuples
+        data class Quadruple<T1, T2, T3, T4>(val t1: T1, val t2: T2, val t3: T3, val t4: T4)
+
+        val (poTokenGenerator, visitorData, streamingPot, hasBeenRecreated) =
+            synchronized(WebPoTokenGenLock) {
+                val shouldRecreate = webPoTokenGenerator == null || forceRecreate ||
+                    webPoTokenGenerator!!.isExpired()
```

Comment:
```suggestion
                val shouldRecreate = forceRecreate || webPoTokenGenerator?.isExpired() == true
```

Reply:
I know this is a possible syntax, but I preferred making the null check clear to have better clarity